airflow/providers/amazon/aws/hooks/base_aws.py

```diff
@@ -754,7 +754,9 @@ def test_connection(self):
         """
         try:
             session = self.get_session()
-            conn_info = session.client("sts").get_caller_identity()
+            conn_info = session.client(
+                "sts", endpoint_url=self.conn_config.endpoint_url
+            ).get_caller_identity()
```

Comment:
In case if custom endpoint set in connection then this connection only would work with STS API, otherwise it will raise an error and vice versa if user provide something like `https://elasticmapreduce.eu-west-1.amazonaws.com` then connection test will fail

Unfortunetly this is a limitation of current implementation: only one endpoint for connection.

So this should be fixed on connection level, for example add ability provide endpoint on service level with some backward compatibility

Reply:
So you think we need to implement something like the following on the connection extra details..

```json
{
    "service_endpoints": {
        "s3": "https://s3.eu-west-2.amazonaws.com",
        "emr": "https://elasticmapreduce.eu-west-1.amazonaws.com"
    },
    "endpoint_url": "http://localstack:4566"
}
```

where the connection should do something like the following... (very pseudo code):

```python

def get_client(client_type):
    
    service_endpoint = self.extra.service_endpoints.get("client_type", self.endpoint_url)
    return self.client(client_type, endpoint_url=service_endpoint
```

where we first attempt to get a service level endpoint url, falling back to a custom/overall endpoint url, which can default to `None` if not set, therefore allowing the default AWS endpoint url to be used?

Reply:
Yep, something like that. With some additional logic most related to STS client and historical reason.
If STS client defined internally in `AwsGenericHook` and `BaseSessionFactory` it should not fallback to default `endpoint_url`. 

Usual users need some additional time for change their connections and if we can not break their connections and workflow it would be nice. Additional logic + warnings it is better approach.

There is two places where we use STS client internally
1. Check credentials by call STS GetCallerIdentity
2. Assume role

https://github.com/apache/airflow/blob/b3e26560c7fd835570a0b3a9d65670c87c8cfe0a/airflow/providers/amazon/aws/hooks/base_aws.py#L184-L186

In case if user use [StsHook](https://github.com/apache/airflow/blob/b3e26560c7fd835570a0b3a9d65670c87c8cfe0a/airflow/providers/amazon/aws/hooks/sts.py#L22) or `AwsBaseHook(client_type="sts")` then we should fallback to default `endpoint_url`. The same how it works now

Reply:
I worked on the same feature, my implementation something like that for `AwsConnectionWrapper` method

```python
    def get_service_endpoint(self, service_name: str, *, sts_use_default: bool = False) -> str | None:
        """
        Return endpoint_url for specific service.

        :param service_name: The name of a service, e.g. 's3' or 'ec2'.
        :param sts_use_default: Is associate default endpoint url from config with sts service.
            For historical reason AwsBaseHook never use endpoint url from
            Connection because this could break some connections, e.g. obtained by Assume Role.
        """
        if not self._endpoint_url:
            return None

        try:
            return self._endpoint_url[service_name]
        except KeyError:
            default = self._endpoint_url.get("_default")
            if service_name != "sts" or sts_use_default:
                return default
            elif default:
                warnings.warn(
                    "Default endpoint URL disabled for 'sts' service, please set it explicit.",
                    UserWarning,
                    stacklevel=3,
                )
            return None
```

And some extending in connection:
1. Allow set `endpoint_url` as dictionary
2. If `endpoint_url` set as non empty string then convert it to `{'_default': endpoint_url}`

This just a sample, I do not have strong opinion about final implementation. 
My only point about the current limitation of `endpoint_url` and why we do not provide it into STS client